hammerblade/torch/kernel/hb_elementwise_for.hpp

```diff
@@ -0,0 +1,370 @@
+//====================================================================
+// Element-wise for helper function
+// 03/12/2020 Lin Cheng (lc873@cornell.edu)
+//====================================================================
+
+#ifndef _HB_ELEMENTWISE_FOR_HPP
+#define _HB_ELEMENTWISE_FOR_HPP
+
+#include <map>
+#include <math.h>
+#include <initializer_list>
+#include <bsg_assert.hpp>
+#include <bsg_tensor.hpp>
+
+// =========================================================
+// Linear index to offset
+// =========================================================
+template<typename scalar_t>
+inline uint32_t offset_calc(uint32_t idx, BSGTensor<scalar_t> tensor) {
+  uint32_t* strides = tensor.get_strides();
+  uint32_t* sizes = tensor.get_sizes();
+  uint32_t factor = 1;
+  uint32_t offset = 0;
+  for(int32_t i = tensor.ndim() - 1; i >= 0; i--) {
+    uint32_t dimx = idx % sizes[i];
+    offset += dimx * strides[i];
+    idx /= factor;
+    factor *= sizes[i];
+  }
+  return offset;
+}
```

Comment:
Can't we just assume tensors in TensorIterators are 1-d tensors (in terms of words)? That saves us lot of index calculation, I believe.

Reply:
Well, actually no. If a tensor is contiguous in memory, then it can be viewed as 1-d tensor. We handle it with the trivial 1d case here
https://github.com/cornell-brg/hb-pytorch/blob/f52ffeaf665e064fbe42a5513bc6d88717cf72a8/hammerblade/torch/kernel/hb_elementwise_for.hpp#L63-L70
here we can avoid any offset calculation

the problem is tensors that are non-contiguous in memory. Originally iter.serial_foreach takes care of splitting this kind of tensors into subiters (which can be handled by trivial 1d case). Now without serial_foreach, we need to handle them by hand.

Reply:
Cool! I imagine, iterator does this for us: If contiguous, generate 1-d tensors?